datumaro/cli/util/project.py

```diff
@@ -22,17 +27,93 @@ def generate_next_file_name(basename, basedir='.', sep='.', ext=''):
 
     return generate_next_name(os.listdir(basedir), basename, sep, ext)
 
-def generate_next_name(names, basename, sep='.', suffix='', default=None):
-    pattern = re.compile(r'%s(?:%s(\d+))?%s' % \
-        tuple(map(re.escape, [basename, sep, suffix])))
-    matches = [match for match in (pattern.match(n) for n in names) if match]
+def parse_full_revpath(s, ctx_project):
+    # named groups cannot be duplicated
+    _counter = count()
+    def maybe_quoted(exp, g=None):
+        if g is None:
+            g = '__tmpg' + str(next(_counter))
+        return rf"(?P<{g}>['\"]?){exp}(?P={g})"
 
-    max_idx = max([cast(match[1], int, 0) for match in matches], default=None)
-    if max_idx is None:
-        if default is not None:
-            idx = sep + str(default)
+    if ctx_project:
+        env = ctx_project.env
+    else:
+        env = Environment()
+
+    def parse_dataset_pathspec(s):
+        regex = "%s(:%s)?" % (
+            maybe_quoted(r"(?P<dataset_path>[^:]+)"),
```

Comment:
This is where `shlex` could be really in its place, because we're trying to pass a path through the command line. Quotes are needed to support spaces in path.

The colons on Windows is a problem, indeed.

Reply:
> This is where shlex could be really in its place, because we're trying to pass a path through the command line. Quotes are needed to support spaces in path.

Can you explain it in more detail? I don't understand why quoting would be necessary to support spaces. If your path is `a b`, and your format is `c`, why couldn't you just use `a b:c` as the pathspec?

Reply:
Hmm, actually right. We type one of
```bash
cmd path/to\ dataset/
cmd path/to\ dataset/:format
cmd "path/to dataset/"
cmd "path/to dataset/:format"
```
and there are no problems with spaces in all the cases. Ok.

Reply:
I added escaping for colons with slashes to support Windows absolute paths.